8032218: Emit single post-constructor barrier for chain of superclass constructors

caojoshua opened the pull request:
[C2 emits a StoreStore barrier for each constructor call](https://github.com/openjdk/jdk/blob/72ca7bafcd49a98c1fe09da72e4e47683f052e9d/src/hotspot/share/opto/parse1.cpp#L1016) in a chain of superclass constructor calls. It is unnecessary. We only need to emit a single barrier for each object allocation / each pair of `Allocation/InitializeNode`. 

[Macro expansion emits a trailing StoreStore after an InitializeNode](https://github.com/openjdk/jdk/blob/32946e1882e9b22c983cbba3c6bda3cc7295946a/src/hotspot/share/opto/macro.cpp#L1610-L1628). This `StoreStore` is sufficient as the post-constructor barrier. From the [InitializeNode definition](https://github.com/openjdk/jdk/blob/32946e1882e9b22c983cbba3c6bda3cc7295946a/src/hotspot/share/opto/memnode.cpp#L3639-L3642):

> // An InitializeNode collects and isolates object initialization after
// an AllocateNode and before the next possible safepoint.  As a
// memory barrier (MemBarNode), it keeps critical stores from drifting
// down past any safepoint or any publication of the allocation.

This PR modifies `Parse::do_exits()` such that it only emits a barrier for a constructor if we find that the constructed object does not have an `InitializeNode`. It is possible that we cannot find an `InitializeNode` i.e. if the outermost method of the compilation unit is the constructor. We still need to emit a barrier in these cases.

Passes hotspot tier1 locally on x86 linux machine. New tests make sure that there is a single `StoreStore` for chained constructors.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8032218](https://bugs.openjdk.org/browse/JDK-8032218): Emit single post-constructor barrier for chain of superclass constructors (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18870/head:pull/18870` \
`$ git checkout pull/18870`

Update a local copy of the PR: \
`$ git checkout pull/18870` \
`$ git pull https://git.openjdk.org/jdk.git pull/18870/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18870`

View PR using the GUI difftool: \
`$ git pr show -t 18870`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18870.diff">https://git.openjdk.org/jdk/pull/18870.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18870#issuecomment-2067359419)

caojoshua commented:
I originally had a [different approach](https://github.com/caojoshua/jdk/commit/eb48719f75279721bcce48d00ed262ebbf3691e0) where we detect superclass constructors and do not emit barriers for them. We can tell if a constructor is a superclass constructor if:

1. the current and parent method have the same receiver
2. the current and parent method are both initializers

I didn't complete it. It fails a case where the superclass constructor does not write a final, and the child constructor does write a final. The patch would not emit a barrier in that case because the outermost constructor did not write a final. The child `Parse` would need to propagate `wrote_final` and other `wrote_*` flags to the parent `Parse`. `Parse` instances don't have any information on each other, and the code would need a fair amount of restructuring to support it.

Despite the bug, it passed all the test cases I created anyway. Thats when I looked into the macro expansion for `InitializeNode` and thought that by itself was sufficient in emitting post-constructor barriers, arriving to this patch.

Sharing this in case I have some wrong assumptions in this PR. We could revert to working o

dean-long commented:
Doesn't this only work if the allocation and call to ctor are compiled together?  Where is the StoreStore added if we compile a <init> method by itself?  The allocation could be in the interpreter but the when calling <init> we call compiled code.

caojoshua commented:
> Doesn't this only work if the allocation and call to ctor are compiled together? Where is the StoreStore added if we compile a method by itself? The allocation could be in the interpreter but the when calling we call compiled code.

Yes, I think you're right. We should probably still emit the barrier when it is the outermost method. Unfortunately, I don't think we can directly IR test on a constructor.

dean-long reviewed (changes requested):
If the subclass ctor inlines the super ctor, it might be safe to omit the barrier from the super ctor.  However, the current approach seem to remove barriers that are needed when the `<init>` is called from the interpter or another compiled method.

caojoshua commented:
Had to step away for a while but still plan on working on this. Comment to keep alive.

caojoshua commented:
Re-worked this patch. See updated description. Now we avoid emitting a barrier if we can find the object's `InitializeNode`. We don't need to emit a barrier if we find the `InitializeNode` because  macro expansion will expand it and emit the trailing StoreStore barrier

dean-long commented on src/hotspot/share/opto/parse1.cpp:
```diff
@@ -1009,12 +1014,12 @@ void Parse::do_exits() {
+      && (wrote_final() || (AlwaysSafeConstructors && wrote_fields()) ||
+        (support_IRIW_for_not_multiple_copy_atomic_cpu && wrote_volatile()))) {
+    _exits.insert_mem_bar(UseStoreStoreForCtor ? Op_MemBarStoreStore :
+        Op_MemBarRelease, alloc_with_final());
```
FYI, JDK-8333791 proposes changes here also.

caojoshua commented:
Had some merge conflicts with https://bugs.openjdk.org/browse/JDK-8333791. As I understand it, when it comes to post-constructor barriers, we can treat stable and final fields the same. Fixed the merge conflict and added tests to reflect those changes.

TobiHartmann commented:
`compiler/stringopts/TestStringObjectInitialization.java` fails on Linux AArch64 with `-XX:-TieredCompilation -XX:+AlwaysIncrementalInline`:

```
java.lang.NullPointerException: Cannot read the array length because "this.value" is null
	at java.base/java.lang.String.length(String.java:1593)
	at java.base/java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:590)
	at java.base/java.lang.StringBuilder.append(StringBuilder.java:179)
	at compiler.stringopts.TestStringObjectInitialization.add(TestStringObjectInitialization.java:62)
	at compiler.stringopts.TestStringObjectInitialization.run(TestStringObjectInitialization.java:67)
	at compiler.stringopts.TestStringObjectInitialization$Runner.run(TestStringObjectInitialization.java:85)
	at java.base/java.lang.Thread.run(Thread.java:1575)
STATUS:Failed.`main' threw exception: java.lang.NullPointerException: Cannot read the array length because "this.value" is null
java.lang.NullPointerException: Cannot read the array length because "this.value" is null
	at java.base/java.lang.String.length(String.java:1593)
	at java.base/java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:590)
	at java.base/java.lang.StringBuilder.append(StringBuilder.java:179)
	at compiler.stringopts.TestStringObjectInitialization.add(TestStringObjectInitialization.java:62)
	at compiler.stringopts.TestStringObjectInitialization.run(TestStringObjectInitialization.java:67)
	at compiler.stringopts.TestStringObjectInitialization$Runner.run(TestStringObjectInitialization.java:85)
	at java.base/java.lang.Thread.run(Thread.java:1575)
STATUS:Failed.`main' threw exception: java.lang.NullPointerException: Cannot read the array length because "this.value" is null
java.lang.NullPointerException: Cannot read the array length because "this.value" is null
	at java.base/java.lang.String.length(String.java:1593)
	at java.base/java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:590)
	at java.base/java.lang.StringBuilder.append(StringBuilder.java:179)
	at compiler.stringopts.TestStringObjectInitialization.add(TestStringObjectInitialization.java:62)
	at compiler.stringopts.TestStringObjectInitialization.run(TestStringObjectInitialization.java:67)
	at compiler.stringopts.TestStringObjectInitialization$Runner.run(TestStringObjectInitialization.java:85)
	at java.base/java.lang.Thread.run(Thread.java:1575)
```

I also see this with an internal stress test:

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/workspace/open/src/hotspot/share/opto/escape.cpp:4674), pid=22930, tid=42355
#  assert(n->is_Mem()) failed: memory node required.
#
# JRE version: Java(TM) SE Runtime Environment (24.0) (fastdebug build 24-internal-2024-08-26-0524112.tobias.hartmann.jdk3)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 24-internal-2024-08-26-0524112.tobias.hartmann.jdk3, mixed mode, tiered, compressed class ptrs, z gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0xbcc316]  ConnectionGraph::split_unique_types(GrowableArray<Node*>&, GrowableArray<ArrayCopyNode*>&, GrowableArray<MergeMemNode*>&, Unique_Node_List&)+0x3796
#


Current CompileTask:
C2:617119 124774       4       sun.util.locale.LocaleExtensions::toID (136 bytes)

Stack: [0x00007fd262f9b000,0x00007fd26309b000],  sp=0x00007fd263095a70,  free space=1002k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0xbcc316]  ConnectionGraph::split_unique_types(GrowableArray<Node*>&, GrowableArray<ArrayCopyNode*>&, GrowableArray<MergeMemNode*>&, Unique_Node_List&)+0x3796  (escape.cpp:4674)
V  [libjvm.so+0xbd46b9]  ConnectionGraph::compute_escape()+0x20e9  (escape.cpp:397)
V  [libjvm.so+0xbd4f11]  ConnectionGraph::do_analysis(Compile*, PhaseIterGVN*)+0xf1  (escape.cpp:119)
V  [libjvm.so+0x9fbeaa]  Compile::Optimize()+0x63a  (compile.cpp:2324)
V  [libjvm.so+0x9ffe13]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x1b43  (compile.cpp:852)
V  [libjvm.so+0x84f2c5]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x1d5  (c2compiler.cpp:142)
V  [libjvm.so+0xa0bad8]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0x928  (compileBroker.cpp:2303)
V  [libjvm.so+0xa0c768]  CompileBroker::compiler_thread_loop()+0x478  (compileBroker.cpp:1961)
V  [libjvm.so+0xeb67bc]  JavaThread::thread_main_inner()+0xcc  (javaThread.cpp:758)
V  [libjvm.so+0x17e2ab6]  Thread::call_run()+0xb6  (thread.cpp:225)
V  [libjvm.so+0x14cc747]  thread_native_entry(Thread*)+0x127  (os_linux.cpp:858)
```
